overlord/snapstate/handlers.go

```diff
@@ -3512,16 +3520,113 @@ func (m *SnapManager) doRemoveAliases(t *state.Task, _ *tomb.Tomb) error {
 	}
 	snapName := snapsup.InstanceName()
 
-	err = m.backend.RemoveSnapAliases(snapName)
-	if err != nil {
+	var removeReason removeAliasesReason
+	if err := t.Get("remove-reason", &removeReason); err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	tr := config.NewTransaction(st)
+	experimentalRefreshAppAwareness, err := features.Flag(tr, features.RefreshAppAwareness)
+	if err != nil && !config.IsNoOption(err) {
 		return err
 	}
 
+	if removeReason != removeAliasesReasonRefresh || !experimentalRefreshAppAwareness || excludeFromRefreshAppAwareness(snapsup.Type) {
+		err = m.backend.RemoveSnapAliases(snapName)
+		if err != nil {
+			return err
+		}
+	} else {
+		t.Set("remove-aliases-skipped", true)
+	}
+
 	snapst.AliasesPending = true
 	Set(st, snapName, snapst)
 	return nil
 }
 
+func (m *SnapManager) undoRemoveAliases(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.InstanceName()
+	curAliases := snapst.Aliases
+
+	var removeAliasesSkipped bool
+	if err := t.Get("remove-aliases-skipped", &removeAliasesSkipped); err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	if !removeAliasesSkipped {
+		_, _, err = applyAliasesChange(snapName, autoDis, nil, snapst.AutoAliasesDisabled, curAliases, m.backend, doApply)
```

Comment:
Great catch, thank you! `undoSetupAliases` definitely needs updating to match expectations.

we were not missing tests really, because before this PR `doSetupAliases` and `doRemoveAliases` were each other's `undo` because they were very simple and straightforward so it made sense then.

I will add unit tests to this PR to account for this undo behavior as mentioned in the PR description.

Reply:
I updated the PR, thanks for pointing this out. I went with reusing the generic `undoRefreshAliases`.